rust/datafusion/src/optimizer/boolean_comparison.rs

```diff
@@ -0,0 +1,270 @@
+                    },
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::NotEq,
+                        right: Box::new(right),
+                    },
+                },
+                _ => Expr::BinaryExpr {
+                    left: Box::new(left),
+                    op: op.clone(),
+                    right: Box::new(right),
+                },
+            }
+        }
+        Expr::Not(expr) => Expr::Not(Box::new(optimize_expr(&expr))),
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+        } => {
+            if expr.is_none() {
+                // recurse into CASE WHEN condition expressions
+                Expr::Case {
+                    expr: None,
+                    when_then_expr: when_then_expr
+                        .iter()
+                        .map(|(when, then)| (Box::new(optimize_expr(when)), then.clone()))
+                        .collect(),
+                    else_expr: else_expr.clone(),
+                }
+            } else {
+                // when base expression is specified, when_then_expr conditions are literal values
+                // so we can just skip this case
+                e.clone()
+            }
+        }
+        Expr::Alias { .. }
+        | Expr::Negative { .. }
+        | Expr::Column { .. }
+        | Expr::InList { .. }
+        | Expr::IsNotNull { .. }
+        | Expr::IsNull { .. }
+        | Expr::Cast { .. }
+        | Expr::ScalarVariable { .. }
+        | Expr::Between { .. }
+        | Expr::Literal { .. }
+        | Expr::ScalarFunction { .. }
+        | Expr::ScalarUDF { .. }
+        | Expr::AggregateFunction { .. }
+        | Expr::AggregateUDF { .. }
+        | Expr::Sort { .. }
+        | Expr::Wildcard => e.clone(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::logical_plan::{col, lit, LogicalPlanBuilder};
+    use crate::test::*;
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let mut rule = BooleanComparison::new();
+        let optimized_plan = rule.optimize(plan).expect("failed to optimize plan");
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+    }
+
+    #[test]
+    fn simplify_eq_expr() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .filter(col("a").eq(lit(true)))?
+            .filter(col("b").eq(lit(false)))?
+            .project(vec![col("a")])?
+            .build()?;
+
+        let expected = "\
+        Projection: #a\
+        \n  Filter: NOT #b\
+        \n    Filter: #a\
+        \n      TableScan: test projection=None";
+
+        assert_optimized_plan_eq(&plan, expected);
+        Ok(())
+    }
+
+    #[test]
+    fn simplify_not_eq_expr() -> Result<()> {
```

Comment:
One suggestion on the tests (not needed in this PR) would be to separtate out the cases -- 
1. one set in terms of `LogicalPlan` that makes sure rewrite was applied to all expressions in the plan node 
2. Another set for just the logic in `optimize_expr` (aka make an `Expr`, rewrite it, and verify the rewrite was done correctly

Reply:
good suggestion, i will add it in this PR